### PR TITLE
VF2 graph isomorphism revised PR for issue #87, requesting review 

### PR DESCRIPTION
**Pull request by sracha4355**
Changes I've made:
- removed unnecessary functions from .h, .h contains only the function interface that will be used by the end-user
- added comments to my code and tests
- added some new tests to test isomorphism
- formatted the code with clang-format using style Google
I appreciate any constructive criticism you guys can offer. As for the comments, it is hard to tell the full story of the algorithm given it's complex nature, so in addition to my comments I linked the research paper if any devs are still confused. 
Thank you guys!

### CODE REVIEW

**Comment by sracha4355** [comment]
Most of the added lines in PR are from cmake files i forgot to push because I rebuilt the library.

**Comment by sracha4355** [comment]
@bobluppes @joweich could I get a workflow approval?

**Comment by sracha4355** [comment]
@joweich can I get a review

File: include/graaflib/algorithm/graph_isomorphism.tpp [hunk 1]
```diff
@@ -0,0 +1,740 @@
...
+  }
+  for (const auto& vertex : rhs.get_vertices()) {
+    std::size_t number_of_conn = rhs.get_neighbors(vertex.first).size();
+    rhs_node_degrees.push_back(number_of_conn);
+  }
+  std::sort(lhs_node_degrees.begin(), lhs_node_degrees.end());
+  std::sort(rhs_node_degrees.begin(), rhs_node_degrees.end());
+  for (int i = 0; i < lhs_node_degrees.size(); i++)
+    if (lhs_node_degrees[i] != rhs_node_degrees[i]) return false;
+  return true;
```
Comment (joweich) [inline include/graaflib/algorithm/graph_isomorphism.tpp:737]:
```suggestion
  auto compute_and_sort_degrees = [](const graph<V, E, T>& graph) {
      std::vector<std::size_t> node_degrees;
      for (const auto& vertex : graph.get_vertices()) {
          std::size_t number_of_conn = graph.get_neighbors(vertex.first).size();
          node_degrees.push_back(number_of_conn);
      }
      std::sort(node_degrees.begin(), node_degrees.end());
      return node_degrees;
  };

  const auto lhs_node_degrees = compute_and_sort_degrees(lhs);
  const auto rhs_node_degrees = compute_and_sort_degrees(rhs);

  return lhs_node_degrees == rhs_node_degrees;
```
I think using a lambda would be more concise here

File: include/graaflib/algorithm/graph_isomorphism.tpp [hunk 2]
```diff
@@ -0,0 +1,740 @@
...
+
+  if (valid_mapping_found) {
+    vertex_mapping mapping;
+    // map vertex ids from lhs to rhs
+    for (int i = 0; i < state->lhs_core.size(); i++)
+      mapping[state->lhs_reverse_node_order.at(i)] =
+          state->rhs_reverse_node_order.at(state->lhs_core[i]);
+    delete state;
+    std::optional<vertex_mapping> _optional = mapping;
+    return _optional;
```
Comment (joweich) [inline include/graaflib/algorithm/graph_isomorphism.tpp:187]:
```suggestion
    return mapping;
```
This should be possible with implicit template deduction.

File: include/graaflib/algorithm/graph_isomorphism.tpp [hunk 3]
```diff
@@ -0,0 +1,740 @@
...
+  for (const auto& element : set_A) {
+    if (set_B.count(element) > 0) intersection.insert(element);
+  }
+  return std::move(intersection);
+}
+
+// a lookahead rule to see if the current partial vertex_mapping can lead to a
+// complete isomorphic vertex_mapping
+template <typename V, typename E, graph_type T>
+bool r_new(const graph<V, E, T>& lhs, const graph<V, E, T>& rhs,
```
Comment (joweich) [inline include/graaflib/algorithm/graph_isomorphism.tpp:670]: this function is very repetitive and its name should be more expressive

File: include/graaflib/algorithm/graph_isomorphism.tpp [hunk 4]
```diff
@@ -0,0 +1,740 @@
...
+    }
+  }
+  return false;
+}
+
+void update_terminal_set(const node_set& vertices, terminal_set& _terminal_set,
+                         size_t depth, size_t& _terminal_set_length) {
+  for (const auto& node : vertices) {
+    if (_terminal_set[node] == -1) {
+      _terminal_set[node] = int(depth);
```
Comment (joweich) [inline include/graaflib/algorithm/graph_isomorphism.tpp:226]: why is depth defined as size_t when casted to int here?

**Review by joweich — COMMENTED** [review]
Hey @sracha4355, great to see you incorporated the feedback from last time!
I'll be a bit more specific with my feedback this time compared to the guardrails last time :slightly_smiling_face: 

- The logic itself is quite hard to follow because the level of abstraction changes often and some sections are quite repetitive. I think we should put some thought into introducing some object orientation to be more concise about the dependencies. `terminal_set` or `vf2_vertex` would be candidates to consider introducing as structs/classes
- There are several return statements that lack test coverage. There should be tests if the returns are necessary, otherwise we should remove them
- You sometimes allocate memory with "naked" new statements. Most often,  it's better (i.e. more robust against memory leaks) to use smart pointers. Won't go into detail here, but check e.g. [this](https://en.cppreference.com/book/intro/smart_pointers)

I did not review every function, but I hope the bullets above and the inline-comments provide some useful advice already! :slightly_smiling_face:

**Comment by sracha4355** [comment]
@joweich thank you for your comments, I'll take a look at each one and refactor the code!

**Comment by sracha4355** [comment]
Hey guys, I got really busy with school, but now that I am on break I would like to refactor the code and get my code merged. Can we reopen the PR?
